docs/proposals/scaled-prometheus-pipeline/scaled-prometheus-pipeline.md

```diff
@@ -0,0 +1,117 @@
+
+### Improving query times with Object Storage
+
+All metrics go through the controller (which is already scalable and placed behind a load balancer), where they are pushed to the `prometheus-edge-hub` (or pushgateway). The single prometheus server then scrapes from this pushgateway. 
+
+All metrics are stored for 30 days in the Prometheus TSDB storage. This means that all queries have to go through the prometheus server itself, which is the main cause of slow queries.
+
+[Object storage](https://thanos.io/storage.md/) will allow us to only store a few hours of metrics on the server itself (potentially keeping everything in-memory) and then exporting older metrics to object storage elsewhere. For example on an AWS deployment metrics would be stored in S3.
+
+*Options for Object Storage*
+* Deployed on AWS: S3
+* Deployed with openstack: Openstack Swift
+* Neither: Use prometheus server and forgo object storage
+
+### Query-side Implementation
+
+We will deploy multiple Querier components behind a load balancer which are configured to talk to both the prometheus server and the Object storage. This will distribute the compute and I/O load away from the prometheus server to the stateless querier components which can be trivially scaled horizontally to handle increased query loads. Currently, the orchestrator API reaches out directly to the Prometheus server to query metrics data, but now the API will point to the Thanos querier components. Nothing else will change in terms of the API.
+
+### Configuration
+
+We should make this as easy to configure as possible. Some goals include:
+
+* Make this entire setup optional. The single prometheus server works well for small to medium deployments.
+* Configuring thanos to run in Orc8r should require no more than a few values in the helm chart. (e.g. `create_thanos` and `prometheus_servers_count)`
+
+* Since we are moving the prometheus-configmanager to act on the Thanos Ruler component, we can manage prometheus server configuration with Kubernetes and ConfigMaps. This will be used to correctly set scrape targets for the servers with respect to the multiple pushgateways.
+
+### Improving prometheus-edge-hub performance
+
+With the goal of handling 10 million datapoints per minute, the [prometheus-edge-hub](github.com/facebookincubator/prometheus-edge-hub) will probably need some improvements to handle that load. First we'll do benchmark tests on specific AWS hardware options. Then, profile the code to find the bottlenecks and improve them. It's not clear what exactly needs to be done, but I'm confident we can find significant improvements as this component hasn't gone through much optimization yet.
+
+In the end if we can't get enough performance out of a single edge-hub, we will have to investigate scaling this horizontally.
+
+
+## Development Plan
+
+|Step	|Est. Time	|
+|---	|---	|
+|Deploy Thanos locally and experiment with loads to validate query time improvements	|2 wk	|
```

Comment:
you'll probably have to deploy this on AWS yourself otherwise you won't be able to test what kind of impact object storage has on query performance